Remove shipping rate method

ecgan opened the pull request:
### Changes proposed in this Pull Request:

In PR https://github.com/woocommerce/google-listings-and-ads/pull/1422#issuecomment-1100022747, the `shippingRate.method` field has been removed from the API.

In this PR, we are removing the `shippingRate.method` and related `"FLAT_RATE"` code from the front-end UI code.

Notable changes: 

- `SHIPPING_RATE_METHOD.FLAT_RATE` (https://github.com/woocommerce/google-listings-and-ads/pull/1445/files#diff-47fa607a658656385fb9df97cafec8e8aee2f1334c174a8ec3f0b45340c01e9eL29-L31) and related usage has been removed.
- `ShippingRate.method` (https://github.com/woocommerce/google-listings-and-ads/pull/1445/files#diff-bf84ff8b2859999848ed9f591fb092253bab6830f00ec3d2a6691206297b6512L39) and related usage has been removed.
- `ShippingRateGroup.method` (https://github.com/woocommerce/google-listings-and-ads/pull/1445/files#diff-dc45687eaf4ca6c9b9f12d7c97cf53e7ad755d9423842dfa0cbcdccdbf304638L8) and related usage has been removed.
- `'flat_rate'` has been removed.
- I've also added test for `groupShippingRatesByFreeShippingThreshold` function.

### Screenshots:

There are no UI changes in this PR. This changes here are logic changes to make the existing UI work with the API changes made in https://github.com/woocommerce/google-listings-and-ads/pull/1422#issuecomment-1100022747.

### Detailed test instructions:

The impacted area is in the shipping rate section as illustrated in the following screenshot: 

![image](https://user-images.githubusercontent.com/417342/165340561-2d1959f6-1f94-4888-90e1-c4aa144e18e0.png)

#### UI test

Go to Setup MC or Edit Free Listings flow and test the following tasks, they should work as expected, as per existing production behavior.

1. Add a shipping rate with rate 0. There should be a pill showing "Free shipping for all orders". You should not be able to add minimum order for the countries.
2. Add a shipping rate with rate greater than 0. Select "Yes" for offer free shipping. Add minimum order for the countries. Change the shipping rate to 0 and the minimum order for the countries should be removed automatically.
3. If all the shipping rates are rate 0, you should not see "offer free shipping" and "minimum order" cards.
4. "Offer free shipping" card should only show up when you have at least one shipping rate with rate greater than 0.
5. "Minimum order" card should only show up when "Yes" is selected in "offer free shipping" card.
6. In shipping rate, click on the edit button and edit the countries and rate.
7. In shipping rate, click on the edit button and then click on the delete button.
8. In "offer free shipping" card, selecting "No" will remove all the minimum order amount. Selecting "Yes" would require you to re-enter the minimum order amount again.
9. In minimum order card, click on the edit button and edit the countries and amount.
10. In minimum order card, click on the edit button and then click on the delete button.

#### Test shipping rates suggestions in Setup MC

1. Make sure you have some existing WC shipping zones settings. You can refer to issue https://github.com/woocommerce/google-listings-and-ads/issues/1395 for an example settings.
2. Disconnect you google accounts and go through Setup MC flow.
3. When you reach Setup MC Step 3, select "flat shipping rates" option. If you are coming to this shipping rate section for the first time, the shipping rates should be automatically pre-filled and saved based on shipping rates suggestions derived from WC shipping zones settings. You can refer to https://github.com/woocommerce/google-listings-and-ads/pull/1422#issuecomment-1101386144 for the example suggestions API response.
4. Reload the page. The shipping rate settings should be the same as before the page reload.

#### Code check

1. Run `npm run test:js`. All tests should run successfully.
5. There should be no more "flat_rate" in the JS code base.
6. Search for "method" in the code base. The results should not be related to shipping rate's method.

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

>

ecgan committed:
Remove method from ShippingRateGroup.

ecgan committed:
Remove method from shippingRate.

ecgan committed:
Remove SHIPPING_RATE_METHOD constant.

ecgan committed:
Rename groupShippingRatesByMethodCurrencyRate to groupShippingRatesByCurrencyRate.

ecgan committed:
Remove unused typedef import.

ecgan committed:
Remove method field in groupShippingRatesByMethodFreeShippingThreshold.

ecgan committed:
Rename groupShippingRatesByMethodFreeShippingThreshold to groupShippingRatesByFreeShippingThreshold.

ecgan committed:
Remove "method: 'flat_rate'" from tests.

ecgan committed:
Add JSDoc for isNonFreeFlatShippingRate.

ecgan committed:
Rename isNonFreeFlatShippingRate to isNonFreeShippingRate.

ecgan committed:
Update JSDoc comment for isSameShippingRate.

ecgan committed:
Update JSDoc for ShippingRate.

ecgan committed:
Add test for groupShippingRatesByFreeShippingThreshold.

eason9487 commented on js/src/components/shipping-rate-section/estimated-shipping-rates-card/groupShippingRatesByCurrencyRate.js:
```diff
@@ -73,23 +65,22 @@
-		const methodCurrencyRate = `${ method } ${ currency } ${ rate } `;
-		const group = rateGroupMap.get( methodCurrencyRate ) || {
+		const { country, currency, rate } = shippingRate;
+		const currencyRate = `${ currency } ${ rate } `;
```
💅 
If I remember right, this variable has been renamed a few times when the grouping rule changes. Considering it's a local variable and used in a small scope, it could be named with another semantic name, such as `key`, rather than naming it according to the grouping values.

ecgan replied:
Addressed in [`5445fd1` (#1445)](https://github.com/woocommerce/google-listings-and-ads/pull/1445/commits/5445fd1ce9816d608d050b610a89af6bb154d8a4). I have also renamed the `rateGroupMap` into `map`. 🙂

eason9487 commented on js/src/components/shipping-rate-section/minimum-order-card/groupShippingRatesByFreeShippingThreshold.js:
```diff
@@ -0,0 +1,36 @@
+ * @param {Array<ShippingRate>} shippingRates Array of shipping rates.
+ * @return {Array<MinimumOrderGroup>} Array of minimum order groups.
+ */
+const groupShippingRatesByFreeShippingThreshold = ( shippingRates ) => {
```
❔ 
The naming convention of this function is a bit different from `groupShippingRatesByCurrencyRate`. The `currency` is one of the grouping factors for them but only one has `currency` within the naming and JSDoc.

ecgan replied:
Addressed in [`e6a0e81` (#1445)](https://github.com/woocommerce/google-listings-and-ads/pull/1445/commits/e6a0e81f759d40b216004b30ff396fbb45d03689).

eason9487 commented on js/src/components/shipping-rate-section/minimum-order-card/groupShippingRatesByFreeShippingThreshold.test.js:
```diff
@@ -0,0 +1,66 @@
+import groupShippingRatesByFreeShippingThreshold from './groupShippingRatesByFreeShippingThreshold';
+
+describe( 'groupShippingRatesByFreeShippingThreshold', () => {
+	it( 'should group the shipping rates based on currency and rate', () => {
```
❓ 
The testing code of this case doesn't cover the "group by currency" part it describes.

ecgan replied:
Addressed in [`09af3a8` (#1445)](https://github.com/woocommerce/google-listings-and-ads/pull/1445/commits/09af3a800f069cd21914ecda061d9f9393dd80ae).

Note that I intentionally added a new test case and mentioned "rare edge case with different currencies" in the test description, so that it is separated from the common expected case.

eason9487 commented on js/src/components/shipping-rate-section/minimum-order-card/groupShippingRatesByFreeShippingThreshold.js:
```diff
@@ -0,0 +1,36 @@
+			currency,
+		} = shippingRate;
+
+		const thresholdCurrency = `${ threshold } ${ currency }`;
```
💅  Same as https://github.com/woocommerce/google-listings-and-ads/pull/1445#discussion_r879024509

ecgan replied:
Addressed in [`36298a2` (#1445)](https://github.com/woocommerce/google-listings-and-ads/pull/1445/commits/36298a2f109982916bac3bd219133efe4384c252).

ecgan committed:
Rename groupShippingRatesByFreeShippingThreshold to groupShippingRatesByCurrencyFreeShippingThreshold.

This is to address https://github.com/woocommerce/google-listings-and-ads/pull/1445#discussion_r879033386.

ecgan committed:
Add test to cover "group by currency" part.

This is to address https://github.com/woocommerce/google-listings-and-ads/pull/1445#discussion_r879036011.

ecgan committed:
Simplify JSDoc.

The usage example is not needed anymore because we have the test file for it now.

ecgan committed:
Simplify variable naming in groupShippingRatesByCurrencyRate.

ecgan committed:
Simplify variable naming in groupShippingRatesByCurrencyFreeShippingThreshold.

eason9487 reviewed (approved):
Thanks for the adjustments! Testing well. LGTM.
